client/lib/logmein/index.ts

```diff
@@ -0,0 +1,121 @@
+/**
+ * Internal Dependencies
+ */
+import getSitesItems from 'calypso/state/selectors/get-sites-items';
+import userUtils from 'calypso/lib/user/utils';
+import { isEnabled } from '@automattic/calypso-config';
+
+/**
+ * Append logmein=direct query parameter to mapped domain urls.
+ *
+ * This triggers a remote-login redirect flow that sets authentication cookies on the
+ * mapped domain enabling the nav bar and other features.
+ */
+
+let reduxStore: any = null;
+
+// Used as placeholder / default domain to detect when we're looking at a relative url
+const INVALID_URL = `https://__domain__.invalid`;
+
+type Host = string;
+
+export function logmeinUrl( url: string ): string {
+	let newurl: URL;
+
+	// Disable feature if not enabled
+	if ( ! isEnabled( 'logmein' ) ) {
+		return url;
+	}
+
+	// logmein=direct only works for logged into wordpress.com users
+	if ( ! userUtils.isLoggedIn() ) {
+		return url;
+	}
+
+	try {
+		newurl = new URL( String( url ), INVALID_URL );
+	} catch ( e ) {
+		// Ignore unparseable urls
+		return url;
+	}
+
+	// Ignore and passthrough /relative/urls that have no host specified
+	if ( newurl.origin === INVALID_URL ) {
+		return url;
+	}
+
+	const sites = Object.values( getSitesItems( reduxStore.getState() ) );
+
+	// Replace unmapped url usage with the mapped hostname
+	const hostmap = unmappedToMapped( sites );
```

Comment:
This creates the `hostmap` on every `logmeinUrl` call, which is inefficient. I think the simplest `sites.find( site => ... )` will be hard to beat in practice.

Reply:
And why do you replace `*.wordpress.com` links with the mapped domain? Keeping them `*.wordpress.com` should be OK.

Reply:
Updating to use the .find method instead.

---

*.wordpress.com urls are used in the preview window and potentially other places. The code in these places is relying on the fact the mapped domains will do a server side redirect. This redirect won't have ?logmein attached so the user will end up on the page potentially logged out when we really want them to be logged in.

It's better if we fix these urls to use the mapped domain directly but I'm not sure how to track all usages down.

Reply:
> And why do you replace *.wordpress.com links with the mapped domain? Keeping them *.wordpress.com should be OK.

Unresolving this as I think I need to justify this a bit better there could be good reasons for linking to the unmapped domain instead of the mapped domain - the domains list for example should rely on server side redirects instead of client side rewrites.

Justification needed is to figure out how many places we do this and if we can't just update them to use the primary domain instead of the unmapped url.

Reply:
Removed this.